In [2]:
import pandas as pd
import matplotlib.pyplot as plt


In [7]:
df= pd.read_csv('/home/protein/Documents/quant_bio/qbb2020-answers/day4-lunch/processed_protein_gene.gtf', header = None, sep = '\t')
df

0        1     2         3         4  5  6  7  \
0     3R  FlyBase  gene    835381   2503907  .  +  .   
1     3R  FlyBase  gene   2554124   3263573  .  -  .   
2     3R  FlyBase  gene   3322810   3354486  .  +  .   
3     3R  FlyBase  gene   3461351   3587054  .  -  .   
4     3R  FlyBase  gene   3623143   3627690  .  +  .   
5     3R  FlyBase  gene   3734855   3745306  .  +  .   
6     3R  FlyBase  gene   3750363   3752353  .  +  .   
7     3R  FlyBase  gene   3846729   3868401  .  +  .   
8     3R  FlyBase  gene   4077939   4104326  .  +  .   
9     3R  FlyBase  gene   4126442   4137882  .  +  .   
10    3R  FlyBase  gene   4160610   4162235  .  -  .   
11    3R  FlyBase  gene   4174632   4186751  .  +  .   
12    3R  FlyBase  gene   4189692   4190260  .  -  .   
13    3R  FlyBase  gene   4197206   4204572  .  +  .   
14    3R  FlyBase  gene   4204493   4215314  .  -  .   
15    3R  FlyBase  gene   4211783   4227518  .  +  .   
16    3R  FlyBase  gene   4218463   4220130  .  -  .   
17    3R  FlyBase  gene   4227387   4229238  .  -  .   
18    3R  FlyBase  gene   4230779   4232342  .  -  .   
19    3R  FlyBase  gene   4233047   4235041  .  +  .   
20    3R  FlyBase  gene   4235217   4241057  .  +  .   
21    3R  FlyBase  gene   4241045   4242717  .  -  .   
22    3R  FlyBase  gene   4242785   4243146  .  -  .   
23    3R  FlyBase  gene   4247020   4248313  .  +  .   
24    3R  FlyBase  gene   4248723   4251341  .  -  .   
25    3R  FlyBase  gene   4251894   4253043  .  -  .   
26    3R  FlyBase  gene   4254708   4255833  .  +  .   
27    3R  FlyBase  gene   4256972   4258444  .  +  .   
28    3R  FlyBase  gene   4266770   4268316  .  +  .   
29    3R  FlyBase  gene   4269221   4277928  .  -  .   
...   ..      ...   ...       ...       ... .. .. ..   
3376  3R  FlyBase  gene  31735503  31736963  .  -  .   
3377  3R  FlyBase  gene  31737035  31737999  .  +  .   
3378  3R  FlyBase  gene  31738468  31743306  .  +  .   
3379  3R  FlyBase  gene  31743490  31744248  .  +  .   
3380  3R  FlyBase  gene  31744282  31745178  .  +  .   
3381  3R  FlyBase  gene  31745172  31746077  .  -  .   
3382  3R  FlyBase  gene  31746445  31747558  .  -  .   
3383  3R  FlyBase  gene  31747643  31751894  .  +  .   
3384  3R  FlyBase  gene  31751825  31754705  .  -  .   
3385  3R  FlyBase  gene  31755356  31758637  .  -  .   
3386  3R  FlyBase  gene  31759404  31761579  .  -  .   
3387  3R  FlyBase  gene  31763304  31764772  .  -  .   
3388  3R  FlyBase  gene  31769293  31771848  .  -  .   
3389  3R  FlyBase  gene  31772219  31775163  .  +  .   
3390  3R  FlyBase  gene  31775049  31776652  .  -  .   
3391  3R  FlyBase  gene  31777020  31790681  .  +  .   
3392  3R  FlyBase  gene  31790592  31793907  .  -  .   
3393  3R  FlyBase  gene  31793981  31799799  .  +  .   
3394  3R  FlyBase  gene  31799634  31805278  .  -  .   
3395  3R  FlyBase  gene  31800963  31802774  .  +  .   
3396  3R  FlyBase  gene  31805468  31806769  .  +  .   
3397  3R  FlyBase  gene  31806867  31810186  .  -  .   
3398  3R  FlyBase  gene  31810460  31812296  .  -  .   
3399  3R  FlyBase  gene  31812609  31846570  .  +  .   
3400  3R  FlyBase  gene  31846987  32015520  .  -  .   
3401  3R  FlyBase  gene  31978121  31990909  .  +  .   
3402  3R  FlyBase  gene  32038330  32039753  .  +  .   
3403  3R  FlyBase  gene  32045314  32051993  .  -  .   
3404  3R  FlyBase  gene  32052060  32054974  .  +  .   
3405  3R  FlyBase  gene  32055279  32068441  .  -  .   

                                                      8  
0     gene_id "FBgn0267431"; gene_version "1"; gene_...  
1     gene_id "FBgn0267430"; gene_version "1"; gene_...  
2     gene_id "FBgn0086917"; gene_version "1"; gene_...  
3     gene_id "FBgn0010247"; gene_version "1"; gene_...  
4     gene_id "FBgn0086378"; gene_version "1"; gene_...  
5     gene_id "FBgn0263977"; gene_version "1"; gene_...  
6     gene_id "FBgn0069923"; gene_version "1"; gene_...  
7     gene_id "FBgn0039955"; gene_version "1"; gene_...  
8     gene_i

In [8]:
columnLbl = ['Seqname', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'Attribute']
df.columns = columnLbl
df

Seqname   Source Feature     Start       End Score Strand Frame  \
0         3R  FlyBase    gene    835381   2503907     .      +     .   
1         3R  FlyBase    gene   2554124   3263573     .      -     .   
2         3R  FlyBase    gene   3322810   3354486     .      +     .   
3         3R  FlyBase    gene   3461351   3587054     .      -     .   
4         3R  FlyBase    gene   3623143   3627690     .      +     .   
5         3R  FlyBase    gene   3734855   3745306     .      +     .   
6         3R  FlyBase    gene   3750363   3752353     .      +     .   
7         3R  FlyBase    gene   3846729   3868401     .      +     .   
8         3R  FlyBase    gene   4077939   4104326     .      +     .   
9         3R  FlyBase    gene   4126442   4137882     .      +     .   
10        3R  FlyBase    gene   4160610   4162235     .      -     .   
11        3R  FlyBase    gene   4174632   4186751     .      +     .   
12        3R  FlyBase    gene   4189692   4190260     .      -     .   
13        3R  FlyBase    gene   4197206   4204572     .      +     .   
14        3R  FlyBase    gene   4204493   4215314     .      -     .   
15        3R  FlyBase    gene   4211783   4227518     .      +     .   
16        3R  FlyBase    gene   4218463   4220130     .      -     .   
17        3R  FlyBase    gene   4227387   4229238     .      -     .   
18        3R  FlyBase    gene   4230779   4232342     .      -     .   
19        3R  FlyBase    gene   4233047   4235041     .      +     .   
20        3R  FlyBase    gene   4235217   4241057     .      +     .   
21        3R  FlyBase    gene   4241045   4242717     .      -     .   
22        3R  FlyBase    gene   4242785   4243146     .      -     .   
23        3R  FlyBase    gene   4247020   4248313     .      +     .   
24        3R  FlyBase    gene   4248723   4251341     .      -     .   
25        3R  FlyBase    gene   4251894   4253043     .      -     .   
26        3R  FlyBase    gene   4254708   4255833     .      +     .   
27        3R  FlyBase    gene   4256972   4258444     .      +     .   
28        3R  FlyBase    gene   4266770   4268316     .      +     .   
29        3R  FlyBase    gene   4269221   4277928     .      -     .   
...      ...      ...     ...       ...       ...   ...    ...   ...   
3376      3R  FlyBase    gene  31735503  31736963     .      -     .   
3377      3R  FlyBase    gene  31737035  31737999     .      +     .   
3378      3R  FlyBase    gene  31738468  31743306     .      +     .   
3379      3R  FlyBase    gene  31743490  31744248     .      +     .   
3380      3R  FlyBase    gene  31744282  31745178     .      +     .   
3381      3R  FlyBase    gene  31745172  31746077     .      -     .   
3382      3R  FlyBase    gene  31746445  31747558     .      -     .   
3383      3R  FlyBase    gene  31747643  31751894     .      +     .   
3384      3R  FlyBase    gene  31751825  31754705     .      -     .   
3385      3R  FlyBase    gene  31755356  31758637     .      -     .   
3386      3R  FlyBase    gene  31759404  31761579     .      -     .   
3387      3R  FlyBase    gene  31763304  31764772     .      -     .   
3388      3R  FlyBase    gene  31769293  31771848     .      -     .   
3389      3R  FlyBase    gene  31772219  31775163     .      +     .   
3390      3R  FlyBase    gene  31775049  31776652     .      -     .   
3391      3R  FlyBase    gene  31777020  31790681     .      +     .   
3392      3R  FlyBase    gene  31790592  31793907     .      -     .   
3393      3R  FlyBase    gene  31793981  31799799     .      +     .   
3394      3R  FlyBase    gene  31799634  31805278     .      -     .   
3395      3R  FlyBase    gene  31800963  31802774     .      +     .   
3396      3R  FlyBase    gene  31805468  31806769     .      +     .   
3397      3R  FlyBase    gene  31806867  31810186     .      -     .   
3398      3R  FlyBase    gene  31810460  31812296     .      -     .   
3399      3R  FlyBase    gene  31812609  3184

In [9]:
lstStart = list(df.iloc[:,3])

# lstStart = df.iloc[0]["Start"] can filter by index and then column
closest_gene = []
numIter = 0
index = 0
maxStart = len(lstStart) - 1
minStart = 0
search = 21378950
gap = len(lstStart) - 1
while (gap > 1):
    index = minStart + int((maxStart - minStart)/2)
    if (search > df.iloc[index]["Start"]):
        if (search > df.iloc[index]["End"]):
            minStart = index
        else:
            print("in the gene")
            break
    elif (search < df.iloc[index]["End"]):
        if (search < df.iloc[index]["Start"]):
            maxStart = index
        else:
            print("in the gene")
            break
    else:
        break
        
    gap = maxStart - minStart
    print(index, gap)
    numIter += 1

1702 1703
2553 851
2127 425
1914 213
2020 107
2073 54
2100 27
2113 13
2106 6
2103 3
2104 1


In [10]:
lstStart[2104]

21378977

In [12]:
df.iloc[2104,:]

Seqname                                                     3R
Source                                                 FlyBase
Feature                                                   gene
Start                                                 21378977
End                                                   21381970
Score                                                        .
Strand                                                       +
Frame                                                        .
Attribute    gene_id "FBgn0004110"; gene_version "1"; gene_...
Name: 2104, dtype: object

## Answers to Binary Search Basic Questions

1. The nearest protein coding gene to the position 3R:21,378,950 is: FBgn0004110

2. The gene's linear genomic distance from 3R:21,378,950 is: 21378977 - 21378950 = 27 nucleotides

3. It took 11 iterations to find this answer.


## Find the 20 nearest protein coding genes to 3R:21,378,950

In [15]:
# works as long as the end and start points are ordered
lstClosest = [2104]
closest_gene = []
index = 2104
lstDist = [abs(search - lstStart[2104])]
lstPot = []
print("Closest protein coding genes to 3R:21,378,950")
print("Number in List, Linear Distance, Gene Name")
while(len(lstClosest) < 20):
    dist = 0
    name = ""
    pot1 = min(lstClosest) - 1
    pot2 = max(lstClosest) + 1
    if (abs(df.iloc[pot1]["End"] - search) < abs(df.iloc[pot2]["Start"] - search)):
        lstClosest.append(pot1)
        dist = abs(df.iloc[pot1]["End"] - search)
        name = df.iloc[pot1]["Attribute"].split('\"')[1].split('\"')[0]
        print(pot1,dist,name)
    elif (abs(df.iloc[pot1]["End"] - search) > abs(df.iloc[pot2]["Start"] - search)):
        lstClosest.append(pot2)
        dist = abs(df.iloc[pot2]["Start"] - search)
        name = df.iloc[pot2]["Attribute"].split('\"')[1].split('\"')[0]
        print(pot2,dist,name)
    closest_gene.append((name,dist))

Closest protein coding genes to 3R:21,378,950
Number in List, Linear Distance, Gene Name
2103 5375 FBgn0267652
2102 5375 FBgn0267650
2101 5375 FBgn0267651
2100 9620 FBgn0261844
2105 9930 FBgn0004862
2099 10671 FBgn0261841
2098 9620 FBgn0261840
2097 10671 FBgn0261843
2096 10972 FBgn0261839
2095 9143 FBgn0261845
2094 10671 FBgn0261842
2093 13660 FBgn0266171
2092 13660 FBgn0266170
2091 13462 FBgn0266172
2090 16819 FBgn0261837
2089 16819 FBgn0261838
2106 18831 FBgn0038886
2088 19182 FBgn0266176
2087 19182 FBgn0266175


## Find the nearest gene of any biotype

In [16]:
df_allGenes = pd.read_csv('/home/protein/Documents/quant_bio/qbb2020-answers/day4-lunch/processed_gene.gtf', header = None, sep = '\t')
columnLbl = ['Seqname', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'Attribute']
df_allGenes.columns = columnLbl
df_allGenes

Seqname   Source Feature     Start       End Score Strand Frame  \
0         3R  FlyBase    gene    722370    722621     .      -     .   
1         3R  FlyBase    gene    835381   2503907     .      +     .   
2         3R  FlyBase    gene   1031171   1031354     .      +     .   
3         3R  FlyBase    gene   2156916   2157206     .      -     .   
4         3R  FlyBase    gene   2554124   3263573     .      -     .   
5         3R  FlyBase    gene   2744304   2744800     .      -     .   
6         3R  FlyBase    gene   3322810   3354486     .      +     .   
7         3R  FlyBase    gene   3461351   3587054     .      -     .   
8         3R  FlyBase    gene   3623143   3627690     .      +     .   
9         3R  FlyBase    gene   3734855   3745306     .      +     .   
10        3R  FlyBase    gene   3750363   3752353     .      +     .   
11        3R  FlyBase    gene   3846729   3868401     .      +     .   
12        3R  FlyBase    gene   4077939   4104326     .      +     .   
13        3R  FlyBase    gene   4126442   4137882     .      +     .   
14        3R  FlyBase    gene   4137894   4138652     .      -     .   
15        3R  FlyBase    gene   4160610   4162235     .      -     .   
16        3R  FlyBase    gene   4174632   4186751     .      +     .   
17        3R  FlyBase    gene   4186963   4196210     .      +     .   
18        3R  FlyBase    gene   4189692   4190260     .      -     .   
19        3R  FlyBase    gene   4197206   4204572     .      +     .   
20        3R  FlyBase    gene   4204493   4215314     .      -     .   
21        3R  FlyBase    gene   4211783   4227518     .      +     .   
22        3R  FlyBase    gene   4218463   4220130     .      -     .   
23        3R  FlyBase    gene   4227387   4229238     .      -     .   
24        3R  FlyBase    gene   4230779   4232342     .      -     .   
25        3R  FlyBase    gene   4233047   4235041     .      +     .   
26        3R  FlyBase    gene   4235217   4241057     .      +     .   
27        3R  FlyBase    gene   4241045   4242717     .      -     .   
28        3R  FlyBase    gene   4242785   4243146     .      -     .   
29        3R  FlyBase    gene   4243606   4245511     .      +     .   
...      ...      ...     ...       ...       ...   ...    ...   ...   
4094      3R  FlyBase    gene  31746445  31747558     .      -     .   
4095      3R  FlyBase    gene  31747643  31751894     .      +     .   
4096      3R  FlyBase    gene  31751825  31754705     .      -     .   
4097      3R  ensembl    gene  31753524  31753591     .      -     .   
4098      3R  FlyBase    gene  31755356  31758637     .      -     .   
4099      3R  FlyBase    gene  31758710  31759360     .      +     .   
4100      3R  FlyBase    gene  31759404  31761579     .      -     .   
4101      3R  FlyBase    gene  31763304  31764772     .      -     .   
4102      3R  FlyBase    gene  31767124  31767790     .      -     .   
4103      3R  FlyBase    gene  31767973  31769423     .      +     .   
4104      3R  FlyBase    gene  31769293  31771848     .      -     .   
4105      3R  FlyBase    gene  31772219  31775163     .      +     .   
4106      3R  FlyBase    gene  31775049  31776652     .      -     .   
4107      3R  FlyBase    gene  31777020  31790681     .      +     .   
4108      3R  FlyBase    gene  31790592  31793907     .      -     .   
4109      3R  FlyBase    gene  31793981  31799799     .      +     .   
4110      3R  FlyBase    gene  31799634  31805278     .      -     .   
4111      3R  FlyBase    gene  31800963  31802774     .      +     .   
4112      3R  FlyBase    gene  31805468  31806769     .      +     .   
4113      3R  FlyBase    gene  31806867  31810186     .      -     .   
4114      3R  FlyBase    gene  31810460  31812296     .      -     .   
4115      3R  FlyBase    gene  31812609  31846570     .      +     .   
4116      3R  FlyBase    gene  31846987  32015520     .      -     .   
4117      3R  FlyBase    gene  31978121  3199

In [17]:
closest_gene = []
numIter = 0
index = 0
maxStart = df_allGenes.shape[0] - 1
minStart = 0
search = 21378950
gap = df_allGenes.shape[0] - 1
while (gap > 2):
    index = minStart + int((maxStart - minStart)/2)
    if (search > df_allGenes.iloc[index]["Start"]):
        if (search > df_allGenes.iloc[index]["End"]):
            minStart = index
        else:
            print("in the gene")
            break
    elif (search < df_allGenes.iloc[index]["End"]):
        if (search < df_allGenes.iloc[index]["Start"]):
            maxStart = index
        else:
            print("in the gene")
            break
    else:
        break
        
    gap = maxStart - minStart
    print(gap)
    print(index)
    numIter += 1

2062
2061
1031
3092
515
2576
258
2318
129
2447
65
2511
33
2543
16
2559
8
2551
4
2547
2
2545


In [18]:
df_allGenes.iloc[2545,:]

Seqname                                                     3R
Source                                                 FlyBase
Feature                                                   gene
Start                                                 21378977
End                                                   21381970
Score                                                        .
Strand                                                       +
Frame                                                        .
Attribute    gene_id "FBgn0004110"; gene_version "1"; gene_...
Name: 2545, dtype: object

The name of this gene is: FBgn0004110 and has the type of being a gene. This is actually the same as the previous search which makes this question really troll....

In [19]:
# this recursive binary search is clearly my own work because
# it is much less elegant than the binary search produced elsewhere
# this binary search is lowkey bad..... because it uses one more index position than a normal recursive binary search

def recursive_binary(search, df, low, high, numIter, prevIndex):
    # base cases:
    if (high - low < 2):
        return (prevIndex,numIter)
 
    if (df.iloc[prevIndex]["Start"] < search and df.iloc[prevIndex]["End"] > search):
        return (prevIndex,numIter)
    
    numIter += 1
    index = low + int(abs(high-low)/2)
    # recursive case
    if (df.iloc[index]["Start"] < search):
        return recursive_binary(search,df,index,high,numIter,index)
    elif (df.iloc[index]["End"] > search):
        return recursive_binary(search,df,low,index,numIter,index)
        
i, count = recursive_binary(21378950, df, 0, df.shape[0]-1, 0, 0)

In [20]:
print(i, count)
#gene_data = data.loc[data['feature'] == 'gene']

2104 11
